pkg/distributor/ha_tracker.go

```diff
@@ -476,19 +483,27 @@ func (h *haTracker) updateKVStore(ctx context.Context, userID, cluster, replica
 		var ok bool
 		if desc, ok = in.(*ReplicaDesc); ok && desc.DeletedAt == 0 {
 			// If the entry in KVStore is up-to-date, just stop the loop.
-			if h.withinUpdateTimeout(now, desc.ReceivedAt) ||
-				// If our replica is different, wait until the failover time.
-				desc.Replica != replica && now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+			if h.withinUpdateTimeout(now, desc.ReceivedAt) {
+				return nil, false, nil
+			}
+		}
+		// If our replica is different, wait until the failover time
+		if desc != nil && desc.Replica != replica {
+			if now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+				level.Info(h.logger).Log("msg", "replica differs, but it's too early to failover", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
 				return nil, false, nil
 			}
+			level.Info(h.logger).Log("msg", "replica differs, attempting to update kv", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
 		}
 
 		// Attempt to update KVStore to our timestamp and replica.
 		desc = &ReplicaDesc{
 			Replica:    replica,
 			ReceivedAt: timestamp.FromTime(now),
 			DeletedAt:  0,
+			ElectedAt:  timestamp.FromTime(now),
```

Comment:
The comment on this field says it's the most recent timestamp when this replica was elected as a leader, but in this code we also update it when we just update the `ReceivedAt` timestamp. We should distinguish both cases and only update the electedAt if replica changed.